### PR TITLE
Limit service API DB threadpool [DPP-998]

### PR DESCRIPTION
**Pull request by simonmaxen-da**

### CODE REVIEW

**Comment by simonmaxen-da** [comment]
Results with queue limited to 50

```
ERROR:
  Code: Aborted
  Message:
 The Index Database Connection Threadpool queue size (51) has exceeded the maximum (50).
 The rejected call was com.daml.ledger.api.v1.PackageService/ListPackages.
 Api services metrics are available at daml.index.db.threadpool.connection.api-server.
```

Screenshot 2022-05-30 at 18 17 40

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 1]
```diff
@@ -15,49 +16,83 @@ import org.slf4j.LoggerFactory
...
-  private val completed = metrics.registry.meter(MetricRegistry.name(apiServices, "completed"))
+  private case class InstrumentedCount(name: String, prefix: MetricName) {
+    private val submitted = metrics.registry.meter(MetricRegistry.name(prefix, "submitted"))
+    private val running = metrics.registry.counter(MetricRegistry.name(prefix, "running"))
+    private val completed = metrics.registry.meter(MetricRegistry.name(prefix, "completed"))
+    def queueSize: Long = submitted.getCount - running.getCount - completed.getCount
+  }
+
+  private val apiServices =
+    InstrumentedCount("Api Services", metrics.daml.lapi.threadpool.apiServices)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:30]:
```suggestion
    InstrumentedCount("Api Services Threadpool", metrics.daml.lapi.threadpool.apiServices)
```
For symmetry's sake

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 2]
```diff
@@ -15,49 +16,83 @@ import org.slf4j.LoggerFactory
...
   override def interceptCall[ReqT, RespT](
       call: ServerCall[ReqT, RespT],
       headers: Metadata,
       next: ServerCallHandler[ReqT, RespT],
   ): ServerCall.Listener[ReqT] = {
 
-    val queued = submitted.getCount - running.getCount - completed.getCount
+    serviceOverloaded(call.getMethodDescriptor.getFullMethodName) match {
+      case Some(errorMessage) =>
+        val rpcStatus = com.google.rpc.Status
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:44]: This error is generated outside of the self-service error code framework. 
It needs to be converted to be compliant, it seems to me `ParticipantBackpressure.Rejection` fits the bill. In this way we will get the same automatic documentation and categorization that is already shared by all the other errors.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/DbDispatcher.scala [hunk 3]
```diff
@@ -119,7 +119,10 @@ object DbDispatcher {
         Some(metrics.registry),
       )
       connectionProvider <- DataSourceConnectionProvider.owner(hikariDataSource)
-      threadPoolName = s"daml.index.db.threadpool.connection.${serverRole.threadPoolSuffix}"
+      threadPoolName = MetricName(
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/DbDispatcher.scala:122]: oops we were defining metrics name outside of the `Metrics.scala`, nice find!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 2 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:44]: Self service error codes are due to be done as part of https://digitalasset.atlassian.net/browse/DPP-1000

**Review by mziolekda — APPROVED** [review]
LGTM

**Pull request merged**